### PR TITLE
Assignment #6

### PR DESCRIPTION
**Pull request by kris1319**
## **Problem.** 

I tried to solve a problem of finding algorithm that contains the minimum possible actions for rearranging. So here it is :)

## **Algorithm.** 

The main idea is to consider arrangements as the permutations.
Assume that we have two arrangements - current [a1, a2, ..., aN] and new [b1, b2, ..., bN], thus let's define:

* transition is a connection between 2 nodes (or 2 cars) v and u such are:
       current[v_new_idx] = u, where v_new_idx - indice of the vth car in the new arrangement;

* chain is a sequence of transitions [v1 -> v2 -> ... -> vK -> 0];

* cycle is a sequence of jumps [v1 -> v2 -> ... -> vK -> v1];

where v1, ..., vK - different nodes (i.e. cars), and 0 is an empty slot. For example:
    
* if [1 2 0] - current, [0 1 2] - new then 1 -> 2 -> 0 is a chain
    
* if [1 2 3 0] - current, [2 3 1 0] - new then 1 -> 2 -> 3 -> 1 is a cycle
    
Also let's notice important facts: 
If we have

* a cycle with size K this cycle can be rearranged by K + 1 steps (simple actions),

* a chain with size K it needs K steps to rearrange the chain.

And these numbers of actions are the only possible minimums. In this way the minimum possible number of steps for rearranging N cars is N.
    
Also it is important that any possible arrangement can be factorized on the set of chains and cycles even if the size of chain/cycle is 1.
The main idea of the method is to traverse two given arrangements (current and new) by chains and
cycles and recursevely rearrange each chain and each cycle.
    
## **Complexity.**
On each chain/cycle the algorithm call itself recursevely and after move current node (car) to the empty slot where the car shoud be in the new arrangement.
One movement contains deleting busy slot and insert new one into the unordered_set of empty slots.
Thus the complexity of one movement is constant - O(1). Assume that the number of cycles is K.
Thus it takes O(N + K) to rearrange the cars.

### CODE REVIEW

File: Assignment_6/arrangement.h [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    ~CarParking() {}
+
+    // Updates current state with the given ordering. It works in assumption that new_arrangement
+    // vector is valid and it is possible to rearrange cars to get a new ordering.
+    //
+    // Algorithm. The main idea is to consider arrangements as the permutations.
+    // Assume that we have two arrangements - current [a1, a2, ..., aN] and new [b1, b2, ..., bN], thus:
+    //
+    // * transition is a connection between 2 nodes (or 2 cars) v and u such are:
+    //   current[v_new_idx] = u, where v_new_idx - indice of the vth car in the new arrangement;
```
Comment (julia57) [inline Assignment_6/arrangement.h]: I think you ment index:)

**Comment by kris1319** [comment]
Yes, my English bugs... :)
Thanks :)

File: Assignment_6/arrangement.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    ~CarParking() {}
+
+    // Updates current state with the given ordering. It works in assumption that new_arrangement
+    // vector is valid and it is possible to rearrange cars to get a new ordering.
+    //
+    // Algorithm. The main idea is to consider arrangements as the permutations.
+    // Assume that we have two arrangements - current [a1, a2, ..., aN] and new [b1, b2, ..., bN], thus:
+    //
+    // * transition is a connection between 2 nodes (or 2 cars) v and u such are:
```
Comment (agaga23) [inline Assignment_6/arrangement.h]: I would say that -> which you use later means transition. And maybe define it without indexes: Transition between x → y is when y stays now (in current arrangment) on the place where we want to put x in new arrangement.

File: Assignment_6/arrangement.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    //
+    // * if [1 2 0] - current, [0 1 2] - new then 1 -> 2 -> 0 is a chain
+    //
+    // * if [1 2 3 0] - current, [2 3 1 0] - new then 1 -> 2 -> 3 -> 1 is a cycle
+    //
+    // Also let's notice important facts: if we have
+    // * a cycle with size K this cycle can be rearranged by K + 1 steps (simple actions),
+    // * a chain with size K it needs K steps to rearrange the chain.
+    // And these numbers of actions are the only possible minimums.
+    // In this way the minimum possible number of steps for rearranging N cars is N.
```
Comment (agaga23) [inline Assignment_6/arrangement.h]: The minimum possible number of steps means that optimal solutions does so. I would change it for something like "In this way we need to do at least N steps for rearranging N cars. More cycles, more steps, but never less." 

Is jump = transition?

The chain size may be also confusing a little bit. Size K means that there are K cars + 0. I would say that size is K+1 in such situation. But maybe it is only my uncertainty.

File: Assignment_6/arrangement.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    //
+    // Also it is important that any possible arrangement can be factorized on the set of chains and
+    // cycles even if the size of chain/cycle is 1.
+    // The main idea of the method is to traverse two given arrangements (current and new) by chains and
+    // cycles and recursively rearrange each chain and each cycle.
+    //
+    // Complexity.
+    // On each chain/cycle the algorithm call itself recursively and after move current node (car) to
+    // the empty slot where the car should be in the new arrangement.
+    // One movement contains deleting busy slot and insert new one into the unordered_set of empty slots.
```
Comment (agaga23) [inline Assignment_6/arrangement.h]: English bug: insert -> inserting. Wouldn't it be better "One movement consist of ..." ? (only suggestion)

File: Assignment_6/arrangement.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    // And these numbers of actions are the only possible minimums.
+    // In this way the minimum possible number of steps for rearranging N cars is N.
+    //
+    // Also it is important that any possible arrangement can be factorized on the set of chains and
+    // cycles even if the size of chain/cycle is 1.
+    // The main idea of the method is to traverse two given arrangements (current and new) by chains and
+    // cycles and recursively rearrange each chain and each cycle.
+    //
+    // Complexity.
+    // On each chain/cycle the algorithm call itself recursively and after move current node (car) to
```
Comment (agaga23) [inline Assignment_6/arrangement.h]: English bug: call -> calls, after move -> moving

File: Assignment_6/arrangement.h [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    // the empty slot where the car should be in the new arrangement.
+    // One movement contains deleting busy slot and insert new one into the unordered_set of empty slots.
+    // Thus the complexity of one movement is constant - O(1). Assume that the number of cycles is K.
+    // Thus it takes O(N + K) to rearrange the cars. So K is less then N thus total complexity is O(N).
```
Comment (agaga23) [inline Assignment_6/arrangement.h]: English bug: then -> than

**Pull request merged**